packages/provider/src/transports/base-provider-transport.ts

```diff
@@ -222,14 +226,17 @@ export abstract class BaseProviderTransport implements ProviderTransport {
       }
     }
 
-    // ACCOUNTS_CHANGED -- when a user logs in or out
+    // ACCOUNTS_CHANGED -- when a user logs in, logs out or disconnects an app
     if (message.type === EventType.ACCOUNTS_CHANGED) {
-      this.accountsChangedPayload = undefined
-      if (message.data && message.data.length > 0) {
-        this.accountsChangedPayload = ethers.utils.getAddress(message.data[0])
-        this.events.emit('accountsChanged', [this.accountsChangedPayload])
+      this.accountsChangedPayload = { accounts: [] }
+      if (message.data && (message.data.length > 0 || message.data.accounts?.length > 0)) {
+        this.accountsChangedPayload = {
+          accounts: [ethers.utils.getAddress(message.data[0] || message.data.accounts[0])],
```

Comment:
It would be nice if we had better typings that discriminate based on the `message.type`, but I guess that makes more sense in a separate PR.

Reply:
Just did a refactor here and removed that part as it wasn't necessary anymore. Moved origin to ProviderMessage so that it won't pollute the data itself.